
## Classification

## Part 1

In [3]:
import numpy as np
import matplotlib.pyplot as plt
from clyent import color


class KNearestNeighbours:
    def __init__(self, __trainingDataFile, __testDataFile):

        self.trainingDataFile = __trainingDataFile
        self.testDataFile = __testDataFile
        self.trainingData = np.array([])
        self.testData = np.array([])
        self.euclideanDistances = np.array([])
        self.euclideanDistancesSorted = np.array([])
        self.readInstances()

    def readInstances(self):

        self.trainingData = np.genfromtxt(self.trainingDataFile, delimiter=',', dtype=float)
        self.testData = np.genfromtxt(self.testDataFile, delimiter=',', dtype=float)

    def viewData(self):
        self.updateDistances()

        
    '''This methods recieves all the training instances, Returns the distance between the single point query
    every training instance. Also return the distances in the sorted manner'''    
    def calculateDistances(self, trainingInstancesMatrix, singlQueryPoint):
        data = np.sqrt(np.sum((trainingInstancesMatrix - singlQueryPoint) ** 2, axis=1))
        sorted_data = np.argsort(data)
        return data, sorted_data

    def updateDistances(self):
        distances = []
        sortedDistances = []
        for i in range(0, np.shape(self.testData)[0]):
            d, sd = self.calculateDistances(self.trainingData[:, :-1], self.testData[i, :-1])
            distances.append(d)
            sortedDistances.append(sd)
            i += 1
        self.euclideanDistances = np.array(distances)
        self.euclideanDistancesSorted = np.array(sortedDistances)

    def CountFrequency(self, my_list):
        '''Returns the value of the majority neighbours '''

        freq = {}
        for item in my_list:
            if (item in freq):
                freq[item] += 1
            else:
                freq[item] = 1
        i = 0
        max_count = 0
        for key, value in freq.items():
            if value > i:
                i = value
                max_count = key
        return max_count

    def calculate_accuracy(self, expected_results, predicted_results):
        indexes = np.where(expected_results == predicted_results)    
        good_pred_count = len(indexes[0])
        accuracy = (good_pred_count / np.shape(kNearestNeighbours.testData)[0]) * 100
        return accuracy




kNearestNeighbours = KNearestNeighbours("trainingData.csv", "testData.csv")
kNearestNeighbours.viewData()

# Select the number of neighbours
k = 1

'''Find the euclidean distance from each instance in test data and find the nearest train instances'''


predicted_results = []
predicted_results_based_on_weightedDistance = []
for indexToBe_searched in range(np.shape(kNearestNeighbours.testData)[0]):
    list_of_neighbours = kNearestNeighbours.euclideanDistancesSorted[indexToBe_searched][:k]

    # print('Here is the list of closest indexes:',lis_of_closest_index)
    results = []
    euclidean_distance_of_neighbours = []
    for neighbour in list_of_neighbours:
        result = kNearestNeighbours.trainingData[neighbour][-1:]

        '''Find the distances between the target instance and the neighbours for weighted distance calculations'''
        euclidean_distance_of_neighbour = kNearestNeighbours.euclideanDistances[indexToBe_searched][neighbour]

        '''Find the result for each neighbour'''
        results.append(float(result))

        euclidean_distance_of_neighbours.append(float(euclidean_distance_of_neighbour))

    '''predict the result based on majority'''
    predicted_results.append([kNearestNeighbours.CountFrequency(results)])

'''Append the obtained prediction results to the test data for convenience'''
final_test_result = np.append(kNearestNeighbours.testData, predicted_results, axis=1)

'''Calculate Accuracy between labelled and predicted data'''
accuracy_by_majority = kNearestNeighbours.calculate_accuracy(final_test_result[:, [-2]], final_test_result[:, [-1]])

print('The accuracy for this KNN-Classification for k={} was ==>{}'.format(k,accuracy_by_majority))






The accuracy for this KNN-Classification for k=1 was ==>89.5


* The above code provides an accuracy of 89.5% for k=1